GTScript-Parallel-model.md

```diff
@@ -134,17 +180,233 @@ The computation domain of every statement is extended to ensure that any require
+
+### Policy for off-center writes
+
+The policy is the following:
+
+- If iteration policy is unspecified then write off-center in the vertical direction is forbidden and raise a compile error
+- If iteration policy is not PARALLEL the write off-center is allowed.
+
+In all this discussion we assumed the fields where off-center writes were performed had been allocated with the proper sizes and index spaces to accommodate them.
+
+<details>
+### Simple example
+We introduce the behavior and policies in GTScript by looking at one sample code executed the in three different iterations policies.
+
+```python
+with computation(FORWARD):
+    with interval(start, end):
+        a[0,0,-1] = c
+        b[0,0,1] = 2 * a # b will see old values of a
+
+with computation(BACKWARD):
+    with interval(start, end):
+        a[0,0,-1] = d
+        b[0,0,1] = 2 * a # b will see newly updated values form d (apart from the first k level (end))
+
+with computation(...):
+    with interval(start, end):
+        a[0,0,-1] = c
+        b[0,0,1] = 2 * a # b will have unspecified values depending on iteration order
 ```
+
+This code is translated in pseudo-code as:
+
+```python
+for k in range(start, end):
+    a[i:I,j:J,k-1] = c[i:I,j:J,k]
+    b[i:I,j:J,k+1] = 2 * a[i:I,j:J,k] # b contains 2 times the previous values in a before the assignment of c, since a and b are evaluated at the same k-level
+
+for k in range(end, start):
+    a[i:I,j:J,k-1] = d[i:I,j:J,k]
+    b[i:I,j:J,k+1] = 2 * a[i:I,j:J,k] # b is going to contain the 2*d except for the firt iteration point, max2.
+
+    # The third case is ambiguous and will raise a compilation error
+```
+
+### Another code sample with conditionals:
+
+```python
+with computation(FORWARD):  # Forward computation
+    with interval(start, end):
+        if (x < 0):
+            a[0,0,-1] = 1
+        else
+            a[0,0,1] = 2
+
+        b = a
+```
+
+Which translates to
+
+```python
+for k in range(start, end):
+    mask[i:I, j:J] = x < 0
+    parfor ij:
+        if (mask):
+            a[0,0,-1] = 1
+        else
+            a[0,0,1] = 2 # This will be visible in the same computation
+
+    parfor ij:
+        b = a
+```
+In this case `b` will contains 2 where `x<0`, otherwise the values contained in `a` before the computation started. This is true except for the first iteration point `min1`, for which `b` contains just the previous values contained in `a`.
```

Comment:
where `x<0` **in the k-1 slice**!!!

I think we should disallow off-center writes in `if`s on fields.

Reply:
I agree with Hannes this kind of mix of ifs and off-center writes could lead to tricky behavior, but I don't see yet why it should be disallowed. Anyway, we need to specify completely the examples to show the potential issues. For example in this case, I would write the example in detail like

```python
# assuming computation domain = [ii:II, jj:JJ, kk:KK]  -> BTW, we should switch to this convention to avoid name clashing with the axes names

with computation(FORWARD):
    with interval(1, KK):
        if (x < 0):
            a[0,0,-1] = 1
        else:
            a[0,0,1] = 2

        b = a
```

translates to:

```python
# Parfor
for k in range(1, KK):
    parfor ij:
        mask[i, j] = (x[i, j, k] < 0)
    parfor ij:
        if mask[i,j]:
            a[i, j, k-1] = 1
        else
            a[i, j, k+1] = 2 # This will be visible in the same computation
    parfor ij:
        b[i, j, k] = a[i, j, k]

# a[i,j,k] == 1 if a[i,j,k+1] < 0
# a[i,j,k] == 2 if a[i,j,k-1] >= 0 and a[i,j,k+1] >= 0
# a[i,j,k] == a_original[i,j,k] in any other point
#
# b[i,j,k] == 2 if a[i,j,k-1] >=0 else a_original[i,j,k]

# NumPy
for k in range(1, KK):
    mask[ii:II, jj:JJ] = (x[k] < 0)
    parfor ij:
        a[ii:II, jj:JJ, k-1][mask] = 1
        a[ii:II, jj:JJ, k+1][mask] = 2
    parfor ij:
        b[ii:II, jj:JJ, k] = a[ii:II, jj:JJ, k]
```

Reply:
I don't think there is a semantic problem. It just gives headache. Maybe a question to @ofuhrer: do you see any cases where you have off-center writes in an if where semantics as described above makes sense?

Reply:
I would not find it ambiguous even if we use the same field in the condition that we update, since the order of updates is clear. Now, this can be tricky to understand, but not as hard as the horizontal cases we discussed for long time

Reply:
I'm not so sure that there is no ambiguity if you update the field in the condition. What would happen in an example like this one: 

```python
with computation(FORWARD):
    with interval(1, KK):
        if (x[0, 0, -1] < 0):
            x[0, 0, -1] += 10:
            if (x[0, 0, -1] < 0):
                a[0,0,0] = 1
        else:
            x[0, 0, -1] = 5
```
In which cases `a == 1`?

Reply:
I just updated the last example to simplify it, but now I don't see any problem here anymore...

Reply:
I do not see the problem here. I does not give me headache, because we are not accessing the same values concurrently. Now, if this is a problem and we could disallow off-center in conditionals, then it's fine with me.

Reply:
I guess it makes sense. The use case I think is advection where you have some velocity and compute where to put the particle. So actually it's exactly the case that you made, with probably the only difference that usually you would accumulate instead of writing a fixed value.

Reply:
So in case @ofuhrer reads this thread, forget my question, I found the answer...

Reply:
So I think the target use case would be something like this
```
with computation(FORWARD):
    with interval(0, KK):
        if (-1 < vel < 0):
            out[0,0,-1] += in[0,0,0]
        else if(-2 < vel < -1):
            out[0,0,-2] += in[0,0,0]
        elseif (0 < vel < 1):
            out[0,0,1] += in[0,0,0]
        etc...
```

Reply:
The same use-case where run-time offsets would be great...

Reply:
I agree that the above example is _dangerous_. Reading the code intuitively completely conveys a different meaning. I currently can't think of a use-case where this would be needed, but there may be vertical solvers that have such a behavior. But honestly, very hard to understand what's going on in this piece of code!!!